### PR TITLE
compute-client: define `Interpreter` and `Transform` APIs

### PR DESCRIPTION
**Pull request by aalexandrov**
This PR proposes a lightweight API for analyzing and transforming LIR plans.

The API should be used conservatively, as in general we want most transformations to happen on the `MIR ⇒ MIR` level. However, recently we have identified several use cases for analysis and transformation on top of LIR plans in #19250 and #19260.

### Motivation

  * This PR adds a feature that has not yet been specified.

In preparation for providing the implementations proposed in #19250 me and @vmarcos discussed providing a more structured API for `LIR` analysis and `LIR ⇒ LIR` transforms.

This might also be of help @mgree once he starts defining LIR-based cost models. We _should be_ able to express this in terms of `Interpreter` implementations, but we might need to tweak the API so we can compose a list of interpreters `I_1, ..., I_m` that need to be inferred in sequence, where `I_m` should get access to the context maps and input values for `I_l` for `l < m`.

### Tips for reviewer

* The first commit is for the abstract interpretation API and the second for the transformation API.
* The API itself is in an `api.rs` sub-module, but all public items are re-exported at the top level.
* The commits provision one file in each of the two modules for the concrete implementation that @vmarcos will work on next.
* Some open questions:
  * I'm not sure if the `transform` and `interpret` modules should live under `plan` or next to `plan`. For example `explain` is next to `plan`, but rendering is within `plan`.
  * At the moment the API is not recursion safe. We can make it recursion-safe in one of several ways:
      * By using stacker through the `mz_ore` extensions (requires changing the return type to `Result`). :heavy_check_mark:
      * By tracking the recursion depth manually (requires changing the return type to `Result`). :x: 
      * By reverting to an iterative algorithms in the `fold` implementations (this will complicate the code somewhat, see the code in `attribute.rs` for the state machines that need to be introduced). :x:

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

File: src/compute-client/src/plan/interpret/api.rs [hunk 1]
```diff
@@ -0,0 +1,650 @@
...
+use crate::plan::join::JoinPlan;
+use crate::plan::reduce::{KeyValPlan, ReducePlan};
+use crate::plan::threshold::ThresholdPlan;
+use crate::plan::top_k::TopKPlan;
+use crate::plan::{AvailableCollections, GetPlan, Plan};
+
+/// An [abstract interpreter] for [Plan] expressions.
+///
+/// This is an [object algebra] / [tagless final encoding] of the language
+/// defined by [crate::plan::Plan], with the exception of the `Get` and `Let*`
```
Comment (vmarcos) [inline src/compute-client/src/plan/interpret/api.rs:35]: `Get` is included, so the exception is only `Let*`?

**Review by vmarcos — APPROVED** [review]
Thanks, this looks great! I will try to use the API tomorrow to create a `BottomUpTransform` for `must_consolidate` relaxation with an associated `Interpreter` for single-time physical monotonicity.

**Comment by vmarcos** [comment]
>   * At the moment the API is not recursion safe. We can make it recursion-safe in one of several ways:
      * By using stacker through the `mz_ore` extensions (requires changing the return type to `Result`). 
      * By tracking the recursion depth manually (requires changing the return type to `Result`). 
      * By reverting to an iterative algorithms in the `fold` implementations (this will complicate the code somewhat, see the code in `attribute.rs` for the state machines that need to be introduced).

No strong opinions here, but it seems like this framework is going to be used after `lower_dataflow`, which already employs `mz_ore` extensions in `from_mir`. So it is reasonable to expect that the calling code would be prepared to handle a `Result`.

File: src/compute-client/src/plan/interpret/api.rs [hunk 2]
```diff
@@ -0,0 +1,650 @@
...
+    }
+
+    /// An immutable fold (structural recursion) over a [Plan] instance.
+    ///
+    /// Runs an abstract interpreter over the given `expr` in a bottom-up
+    /// manner, keeping the `ctx` field of the enclosing field up to date, and
+    /// returns the final result for the entire `expr`.
+    pub fn apply(&mut self, expr: &Plan<T>) -> I::Domain {
+        use Plan::*;
+        match expr {
```
Comment (mgree) [inline src/compute-client/src/plan/interpret/api.rs:200]: Should this use guarded recursion?

Comment (aalexandrov) [reply src/compute-client/src/plan/interpret/api.rs:200]: Yes, I've listed this as an open question - the issue is that if I use guided recursion the return type needs to change, and in principle I think we have an option here to use stack-based recursion.

File: src/compute-client/src/plan/interpret/api.rs (hunk 1 above)
Comment (aalexandrov) [reply src/compute-client/src/plan/interpret/api.rs:35]: Yes, I forgot to update this comment after I figured out that `Get` needs to be part of the interface.

File: src/compute-client/src/plan/interpret/api.rs (hunk 2 above)
Comment (aalexandrov) [reply src/compute-client/src/plan/interpret/api.rs]: I added guided recursion to the two modules in separate commits for easier reviewing.

**Pull request merged**